### PR TITLE
qa: enhance per-client labelled perf counters test

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/65497

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
Tests passed: https://pulpito.ceph.com/jcollin-2024-03-11_08:54:59-fs:functional-wip-ssharon-test2-2024-03-10-1648-distro-default-smithi/

**Comment by vshankar** [comment]
jenkins retest this please

File: qa/tasks/cephfs/test_admin.py [hunk 1]
```diff
@@ -53,21 +52,32 @@ def get_counters_for(filesystem, client_id):
         # write workload
         self.mount_a.create_n_files("test_dir/test_file", 1000, sync=True)
         with safe_while(sleep=1, tries=30, action=f'wait for counters - {mount_a_id}') as proceed:
-            counters_dump_a = get_counters_for(fs_suffix, mount_a_id)
             while proceed():
-                if counters_dump_a["total_write_ops"] > 0 and counters_dump_a["total_write_size"] > 0:
-                    return True
+                counters_dump_a = get_counters_for(fs_suffix, mount_a_id)
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:56]: Maybe also add checks to verify that the avg IO latencies get reflected?

**Comment by joscollin** [comment]
jenkins test make check

**Review by vshankar — CHANGES_REQUESTED** [review]
Looks fine. Suggest also adding verifying perf counters when new clients connect and some clients disconnect/reconnect.

**Comment by joscollin** [comment]
> Looks fine. Suggest also adding verifying perf counters when new clients connect and some clients disconnect/reconnect.

Do we have such events already tested somewhere? Or need to create a new test?

**Comment by joscollin** [comment]
> > Looks fine. Suggest also adding verifying perf counters when new clients connect and some clients disconnect/reconnect.
> 
> Do we have such events already tested somewhere? Or need to create a new test?

never mind, got it.

**Comment by joscollin** [comment]
> Looks fine. Suggest also adding verifying perf counters when new clients connect and some clients disconnect/reconnect.

@vshankar Added more tests. Please take a look.

**Comment by joscollin** [comment]
Tests passed: https://pulpito.ceph.com/jcollin-2024-03-14_13:02:11-fs:functional-main-distro-default-smithi/

File: qa/suites/fs/functional/tasks/admin.yaml [hunk 2]
```diff
@@ -5,6 +5,7 @@ overrides:
         lockdep: true
     log-ignorelist:
       - missing required features
+      - evicting unresponsive client
```
Comment (vshankar) [inline qa/suites/fs/functional/tasks/admin.yaml:8]: Why is ignoring this required? Requires an explanation in the commit message.

**Review by vshankar — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: qa/suites/fs/functional/tasks/admin.yaml (hunk 2 above)
Comment (joscollin) [reply qa/suites/fs/functional/tasks/admin.yaml:8]: Fixed

Comment (joscollin) [reply qa/suites/fs/functional/tasks/admin.yaml:8]: @vshankar Fixed the commit message.

File: qa/tasks/cephfs/test_admin.py [hunk 3]
```diff
@@ -16,15 +16,72 @@
...
+        self.mount_a.kill()
+        self.mount_a.kill_cleanup()
+        self.mount_b.kill()
+        self.mount_b.kill_cleanup()
+        self.fs.set_joinable()
+        log.info("Waiting for reconnect...")
+        self.fs.wait_for_state("up:reconnect", timeout=MDS_RESTART_GRACE)
+        log.info("Waiting for active...")
+        self.fs.wait_for_state('up:active', timeout=MDS_RESTART_GRACE)
+        self.mount_a.mount_wait(cephfs_name=f'{fs_name}')
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:71]: Are you sure this is reconnecting? kill_cleanup does a unmount_wait(force=True) so the mounts are unmounted. For checking reconnections, you'd need to disrupt the network b/w the client and the MDS (something like failing over the MDS would trigger a reconnect from the client).

Comment (vshankar) [reply qa/tasks/cephfs/test_admin.py:71]: Also, post reconnecting, validate the IO perf counters when the client does IO.

**Review by vshankar — CHANGES_REQUESTED** [review]

File: qa/tasks/cephfs/test_admin.py [hunk 4]
```diff
@@ -16,15 +16,72 @@
...
 
 log = logging.getLogger(__name__)
+MDS_RESTART_GRACE = 60
 
 class TestLabeledPerfCounters(CephFSTestCase):
     CLIENTS_REQUIRED = 2
     MDSS_REQUIRED = 1
 
-    def test_per_client_labeled_perf_counters(self):
+    def test_per_client_labeled_perf_counters_on_client_connect(self):
```
[the remark this answers is not in the document]
Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py:25]: Which already existing test you are referring to?

[the remark this answers is not in the document]
Comment (vshankar) [reply qa/tasks/cephfs/test_admin.py:25]: the initial one that I wrote - that already has new clients connecting, isn't it?

[the remark this answers is not in the document]
Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py:25]: The initial one that you wrote, checks for the metrics from the existing clients given by the setup. But `test_per_client_labeled_perf_counters_on_client_connect` checks the metrics specific for a new client connected. Yes, they both does 'almost' similar thing. If you think this is unnecessary, I'm okay with dropping this test which was originated from [this](https://github.com/ceph/ceph/pull/56052#pullrequestreview-1933550373) review comment.

[the remark this answers is not in the document]
Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py]: Dropped

File: qa/tasks/cephfs/test_admin.py (hunk 3 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py]: Fixed

**Comment by joscollin** [comment]
Rebased

File: qa/tasks/cephfs/test_admin.py [hunk 5]
```diff
@@ -16,23 +16,88 @@
...
+
+        with safe_while(sleep=1, tries=30, action='wait for counters') as proceed:
+            while proceed():
+                dump = self.fs.rank_tell(["counter", "dump"])
+                fs_suffix = dump['mds_client_metrics'][0]['labels']['fs_name']
+                per_client_metrics_key = f"mds_client_metrics-{fs_suffix}"
+                clients = [c["labels"]["client"] for c in dump.get(per_client_metrics_key, {})]
+                if mount_a_id in clients and mount_b_id in clients:
+                    # success, got metrics.
+                    break
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:67]: return True??

File: qa/tasks/cephfs/test_admin.py [hunk 6]
```diff
@@ -16,23 +16,88 @@
...
+        mount_a_id = f'client.{self.mount_a.get_global_id()}'
+        self.mount_a.teardown()
+        with safe_while(sleep=1, tries=30, action=f'wait for counters - {mount_a_id}') as proceed:
+            while proceed():
+                dump = self.fs.rank_tell(["counter", "dump"])
+                per_client_metrics_key = f"mds_client_metrics-{dump['mds_client_metrics'][0]['labels']['fs_name']}"
+                clients = [c["labels"]["client"] for c in dump.get(per_client_metrics_key, {})]
+                if clients and mount_a_id not in clients:
+                    # success, no metrics.
+                    break
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:45]: return True??

File: qa/tasks/cephfs/test_admin.py [hunk 7]
```diff
@@ -16,23 +16,88 @@
...
+        self.mount_a.create_n_files("test_dir/test_file", 1000, sync=True)
+        with safe_while(sleep=1, tries=30, action=f'wait for counters - {mount_a_id}') as proceed:
+            while proceed():
+                counters_dump_a = self._get_counters_for(fs_suffix, mount_a_id)
+                if counters_dump_a["total_write_ops"] > 0 and counters_dump_a["total_write_size"] > 0 and \
+                   counters_dump_a["avg_write_latency"] >= 0 and counters_dump_a["avg_metadata_latency"] >= 0 and  \
+                   counters_dump_a["opened_files"] >= 0 and counters_dump_a["opened_inodes"] > 0 and \
+                   counters_dump_a["cap_hits"] > 0 and counters_dump_a["dentry_lease_hits"] > 0 and \
+                   counters_dump_a["pinned_icaps"] > 0:
+                    break
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:80]: ditto

File: qa/tasks/cephfs/test_admin.py [hunk 8]
```diff
@@ -16,23 +16,88 @@
...
+            self.mount_b.open_background(basename=f'test_dir/test_file_{i}', write=False)
+        with safe_while(sleep=1, tries=30, action=f'wait for counters - {mount_b_id}') as proceed:
+            while proceed():
+                counters_dump_b = self._get_counters_for(fs_suffix, mount_b_id)
+                if counters_dump_b["total_read_ops"] >= 0 and counters_dump_b["total_read_size"] >= 0 and \
+                   counters_dump_b["avg_read_latency"] >= 0 and counters_dump_b["avg_metadata_latency"] >= 0 and \
+                   counters_dump_b["opened_files"] >= 0 and counters_dump_b["opened_inodes"] >= 0 and \
+                   counters_dump_b["cap_hits"] > 0 and counters_dump_a["dentry_lease_hits"] > 0 and \
+                   counters_dump_b["pinned_icaps"] > 0:
+                    break
```
Comment (vshankar) [inline qa/tasks/cephfs/test_admin.py:93]: ditto

**Review by vshankar — COMMENTED** [review]
Otherwise LGTM

File: qa/tasks/cephfs/test_admin.py (hunk 6 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py]: Fixed

File: qa/tasks/cephfs/test_admin.py (hunk 5 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py:67]: I've checked the logs during teuthology testing and a `return True` from here doesn't allow the test to continue.

File: qa/tasks/cephfs/test_admin.py (hunk 7 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py:80]: I've checked the logs during teuthology testing and a `return True` from here doesn't allow the test to continue.

File: qa/tasks/cephfs/test_admin.py (hunk 8 above)
Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py:93]: I've checked the logs during teuthology testing and a `return True` from here doesn't allow the test to continue.

File: qa/tasks/cephfs/test_admin.py (hunk 5 above)
Comment (vshankar) [reply qa/tasks/cephfs/test_admin.py:67]: Really? Can you share the test run please?

Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py:67]: I don't have that run with me now. I need to reproduce that issue.

Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py:67]: @vshankar 
See this run: https://pulpito.ceph.com/jcollin-2024-05-02_08:07:52-fs:functional-main-distro-default-smithi/7685184/
I've updated the code with `return True` as you've suggested and `test_per_client_labeled_perf_counters_on_client_reconnect` doesn't continue. It stops after the first `safe_while` (at Line#67).

With `break` the test continues and passes in this run:
https://pulpito.ceph.com/jcollin-2024-05-02_09:55:53-fs:functional-main-distro-default-smithi/7685188/

**Comment by joscollin** [comment]
jenkins test make check

**Comment by vshankar** [comment]
jenkins retest this please

File: qa/tasks/cephfs/test_admin.py (hunk 5 above)
Comment (vshankar) [reply qa/tasks/cephfs/test_admin.py:67]: I see what you mean -- the `safe_while` is nested (its confusing btw).

Just add a comment of why `break` is used.

Comment (joscollin) [reply qa/tasks/cephfs/test_admin.py]: updated

**Review by vshankar — APPROVED** [review]

**Comment by joscollin** [comment]
jenkins test make check

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2024-05-07_03:44:24-fs-wip-vshankar-testing-20240506.153513-testing-default-smithi/

(unfortunately, failed are infra issues related to cephadm - would need a rebuild)

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/65882](https://tracker.ceph.com/issues/65882).

**Comment by vshankar** [comment]
None of the tests in this change ran since the job ran into https://tracker.ceph.com/issues/65841 (fix for which is under dev/test). I'm afraid, this change has to be retested :/

**Comment by joscollin** [comment]
> None of the tests in this change ran since the job ran into https://tracker.ceph.com/issues/65841 (fix for which is under dev/test). I'm afraid, this change has to be retested :/

@vshankar
Please retest it in your another test branch. It should be the issue with https://tracker.ceph.com/issues/65841, as it passed in my tests.

**Comment by joscollin** [comment]
rebased

**Comment by joscollin** [comment]
jenkins test windows

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66327](https://tracker.ceph.com/issues/66327).

**Comment by vshankar** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66522](https://tracker.ceph.com/issues/66522).

**Comment by joscollin** [comment]
jenkins test windows

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-vshankar-testing-20240617094253-debug

(https://tracker.ceph.com/issues/66522)

**Pull request merged**